src/lib.rs

```diff
@@ -196,6 +196,11 @@ impl PublicKeyShare {
         PEngine::pairing(share.0, hash) == PEngine::pairing((self.0).0, *w)
     }
 
+    /// Derives a child public key share for a given index.
+    pub fn derive_child(&self, index: &[u8]) -> Self {
```

Comment:
Would it be better to make index IntoFr?   eg:
```suggestion
    pub fn derive_child<T: IntoFr>(&self, index: T) -> Self {
```
If we already have an Fr, we want to avoid the conversion to bytes and back I think.

Reply:
I feel it would be misleading to have it as a From method, considering how the other `into_fr` methods work.

eg `Fr::from(3u32).to_bytes()` gives `000000...000003`

The same Fr can be created with `3u64` and `3u8`.

But `index_to_fr([3u8]) != index_to_fr(3u32.to_be_bytes()) != index_to_fr(3u64.to_be_bytes)`

To my mind it would be confusing and misleading to have `index_as_bytes` in the same category as `u8` or `u32`.

Reply:
I think I see what you are saying.  or maybe not...?

I guess I was expecting Fr for symmetry with SecretKeySet::secret_key_share():

> pub fn secret_key_share<T: IntoFr>(&self, i: T)

Is there a reason that doesn't make sense, from an API standpoint?   eg, would we ever need to derive a key with an index bigger than an Fr?

Also, I guess I was thinking that the derive_child param would be an Fr (or IntoFr), and then the Fr's bytes would be passed to index_to_fr().  So the 3u8, 3u32, and 3u64 would each be first converted into the same Fr, thus resulting in same derived key for all 3.

> index_to_fr([3u8]) != index_to_fr(3u32.to_be_bytes()) != index_to_fr(3u64.to_be_bytes)

this means that with the present impl, using byte values of 3u8, 3u32, and 3u64 as param will each give a different key, correct?    So that's maybe a bit surprising?

Reply:
I would say that I'm on the fence myself, so if you have a strong feeling either way, its probably best to go with it, and I will happily use the API.  That said, I will throw some thoughts / considerations for you to chew on.

1. We want things to be efficient.  So if extra work is required to convert to Fr first than using bytes directly, then that seems to favor accepting bytes.  If no extra work is required, maybe that favors accepting Fr?

2. Conceptually is this any different from ::secret_key_share(i)?   In my head, they are doing same/similar thing, so the parameter type should be the same.  Is that wrong-thinking?   Otherwise, if Fr is too limiting for ::derive_child(), maybe it is too limiting for ::secret_key_share() also, and it should likewise accept bytes?
 
3. Perhaps due to bitcoin key derivation and also ::secret_key_share(i), when I think of deriving keys I think of integers/indexes.  It's a little weird/confusing to think of byte arrays instead, or that I need to convert my index 1 to 1.to_be_bytes().    just speaking as an API user/consumer.

4. Is there some kind of ToBytes trait that the std integer types implement?  If so, could possibly accept types that impl that trait.  So that ::derive_child(1u32) would "just work".  might just be wishful thinking though.

5. Definitely nice to be able to use 2^256 numbers, or even larger.  It sounds like that can be done with your hashing rounds, but maybe we would need to add Fr::from(&[u8]) or something like that?   So caller could use ::derive_child(Fr::from(5u256.to_be_bytes())) or ::derive_child(2u8).

Reply:
Point 2 has convinced me to change `derive_child` from `[u8]` to `IntoFr` and change `index_to_fr` to `From<[u8]>`

I think even though the details underlying it differ a little, the api has no reason to differ between `derive_child` and `secret_key_share`.

My plan for progress (let me know what you think):

* merge this PR as-is, keeping bytes as the input for all derive_child methods in v3.x.x
* submit a new PR with a breaking change for v4.0.0 that uses IntoFr for all derive_child
* this allows us to have a complete and consistent implementation of derive_child in 3.x.x rather than leaving it unimplemented and going straight to v4.0.0

Reply:
The plan sounds good to me.  :+1: